### ISSUE TITLE
config backup saves user gid but not groupname

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#1781]
As the storageadmin.User Model has a ForeignKey relationship to it’s associated Group, when the django ‘dumpdata’ command is called on the User model, as part of config backup, we end up saving the db id of the associated Group instance. This is a non portable numeric that is db instance specific. Ie the  backup config file contents for a single user entry would be:
```
{
  "fields": {
    "username": "user1",
    "public_key": null,
    "shell": "/bin/bash",
    "group": 29,
    "uid": 1053,
    "admin": false,
    "gid": 1053,
    "user": null,
    "smb_shares": [],
    "email": null,
    "homedir": "/home/user1"
  },
  "model": "storageadmin.user",
  "pk": 42
}
```
where 29 as the “group” value equates to the transient db table row id of Storageadmin_group where groupname = “user1”:

    id        gid        groupname        admin
    29        1053        user1        True

The proposed value here would be the Group.groupname value 'user1' for the linked Group instance. The existing group value has thus far been ignored as there would never have been any match due to its format (numeric) not being a valid group id. Consequently to date a fail over to using the gid contained within the same user config backup record was used. However the preferred, more portable and robust option is to use the group name proper. Hence this issue and associated fix.

The solution found for this issue and developed in concert with #1774 related config restore improvements is:

- [x] Add a natural_key() method to the Group model with a single item return of self.groupname.
- [x] Invoke this natural_key() method on the serialisation of the dumpdata command used to create the config backup file contents.

No regressions were observed re user/group backup and restore when applied after the modifications  to config restore mechanisms associated with issue #1774. It is therefore suggested that this patch be applied after the pr associated with issue #1774 (currently pending).

It is acknowledged that, taken as a whole, the config file contents could be used by cross reference group pk id from a user record to groupname using the storageadmin.group records however this would complicate the restore process and make the existing separation between these processes difficult to maintain. I.e the relevant entry for the example 'user1' group was as follows:
```
{
  "fields": {
    "admin": true,
    "groupname": "user1",
    "gid": 1053
  },
  "model": "storageadmin.group",
  "pk": 29
}
```
Where the pk (primary key) relates to our initially indicated group value.

Attempting to restore users and groups whilst maintaining the same uid and gid of both is impractical and less robust, especially when restoring to a system that has existing users. And so our existing system of giving precedence to group/user names serves well portability wise. And replacing the group pk entry on user records for the group name allows for simpler restore code and eases the human readability of the config backup file. And given group names are install instance unique we should not see duplication. Hence no information is lost but ease of restore, using the existing in place mechanism is facilitated.

It is also understood that early NFS variants deal only in uid, gid, but we already readily change group id when adding users to groups, ie prioritising name over id. Plus the later NFS variants are not limited to uid,gid.

A pr for this issue will follow shortly.

**Comment by phillxnet** [comment]
Post associated patch the user record within a config backup is as follows:
```
{
  "fields": {
    "username": "user1",
    "public_key": null,
    "shell": "/bin/bash",
    "group": "user1",
    "uid": 1053,
    "admin": false,
    "gid": 1053,
    "user": null,
    "smb_shares": [],
    "email": null,
    "homedir": "/home/user1"
  },
  "model": "storageadmin.user",
  "pk": 42
}
```
Note that in both cases the  '--indent 2' option was employed on dumpdata to ease file content examination. This option is however incompatible with current config file restore mechanisms.

**Comment by phillxnet** [comment]
On further consideration I am not wholly taken by my own arguments re complexity of an alternative approach to that indicated here. As such my intention re this issue is to pursue an alternative solution involving a pre-processor, located in:
src/rockstor/storageadmin/views/config_backup.py:restore_users_groups(ml)
where by the user records group field, within the config backup file, is cross referenced with the group records (via the pk) so that the group name can be substituted.
Once proof of concept code is complete I can assess if such an approach is preferred, if so then I will open an new issue in preference to this one and detail accordingly.

**Comment by phillxnet** [comment]
Having explored an alternative solution to this issue, as presented in:
issue #1783 "ensure restored group membership is name based"
and addressed via:
pr #1784 "ensure restored group membership is name based. Fixes #1783"
I am closing this issue as a duplicate of #1783 .
